### ISSUE TITLE
Performance Issue with Jacobian Computation on First-Order Symbolic Vector

### ISSUE THREAD

**Issue by linjing-lab** [sympy/sympy#30186]
Hello, I recently added `sp.lambdify` to optimtool to accelerate numerical computations. However, while testing second-order algorithms, I found that `res=funcs.jacobian(args)` and `res.jacobian(args)` is extremely slow in `newton_quasi.bfgs` method, so that my tests on classification problems remain limited to first-order methods. I would appreciate it if the SymPy developers could provide an alternative symbolic Hessian matrix implementation. Currently, computing the jacobian on a one-dimensional vector basis appears to be the optimal approach for incrementally calculating the Hessian. I’m raising this issue to report the runtime performance bottleneck encountered during optimtool development. I understand this slowdown stems from the scaling complexity of first-order symbolic vectors. However, the code below is merely a test problem for second-order methods—how can I optimize the Hessian computation to achieve similar speedups for classification tasks? Currently, thanks to the `sp.lambdify` optimization, first-order methods have seen a dramatic performance boost.
```python
import optimtool.unconstrain as ou
from optimtool.base import sp, np

def gen_nn(X_data, y_data, hidden_dims=[], task='classification'):
    n_samples, n_features = X_data.shape
    if task == 'classification':
        if len(y_data.shape) == 1:
            n_outputs = len(np.unique(y_data))
            y_one_hot = np.zeros((n_samples, n_outputs))
            for i, label in enumerate(y_data):
                y_one_hot[i, int(label)] = 1
        else:
            n_outputs = y_data.shape[1]
            y_one_hot = y_data
        y_processed = y_one_hot
    elif task == 'regression':
        if len(y_data.shape) == 1:
            n_outputs = 1
            y_processed = y_data.reshape(-1, 1)
        else:
            n_outputs = y_data.shape[1]
            y_processed = y_data
    else:
        raise ValueError(f"Support classification or regression. not support {task}.")
    params = []
    layer_dims = [n_features] + hidden_dims + [n_outputs]
    for i in range(len(layer_dims)-1):
        prev_dim = layer_dims[i]
        curr_dim = layer_dims[i+1]
        for j in range(curr_dim):
            for k in range(prev_dim):
                params.append(sp.symbols(f'W{i}_{j}_{k}', real=True))
        for j in range(curr_dim):
            params.append(sp.symbols(f'b{i}_{j}', real=True))
    def forward(X_vec):
        idx = 0
        x = X_vec.copy()
        current_dim = n_features
        for i, h_dim in enumerate(hidden_dims):
            W_mat = []
            for j in range(h_dim):
                row = []
                for k in range(current_dim):
                    row.append(params[idx])
                    idx += 1
                W_mat.append(row)
            b_vec = []
            for j in range(h_dim):
                b_vec.append(params[idx])
                idx += 1
            z = []
            for j in range(h_dim):
                sum_val = 0
                for k in range(current_dim):
                    sum_val += W_mat[j][k] * x[k]
                sum_val += b_vec[j]
                z.append(sum_val)# sp.Max(0, sum_val)
            x = z
            current_dim = h_dim
        output_dim = n_outputs
        W_mat = []
        for j in range(output_dim):
            row = []
            for k in range(current_dim):
                row.append(params[idx])
                idx += 1
            W_mat.append(row)
        b_vec = []
        for j in range(output_dim):
            b_vec.append(params[idx])
            idx += 1
        out = []
        for j in range(output_dim):
            sum_val = 0
            for k in range(current_dim):
                sum_val += W_mat[j][k] * x[k]
            sum_val += b_vec[j]
            out.append(sum_val)
        return out
    f_sym = 0
    epsilon = 1e-10
    if task == 'classification':
        for s in range(n_samples):
            logits = forward(X_data[s].tolist())
            y_true = y_processed[s].tolist()
            log_sum_exp = sp.log(sum(sp.exp(l) for l in logits) + epsilon)
            probs = [sp.exp(l - log_sum_exp) for l in logits]
            for c in range(n_outputs):
                f_sym += -y_true[c] * sp.log(probs[c] + epsilon) # Softmax
        f_sym /= n_samples 
    else:  # regression
        predictions = []
        for s in range(n_samples):
            pred = forward(X_data[s].tolist())
            predictions.append(pred)
        for s in range(n_samples):
            for o in range(n_outputs):
                diff = predictions[s][o] - y_processed[s, o]
                f_sym += diff ** 2 # MSE
        f_sym /= (n_samples * n_outputs)
    return f_sym, params

# classification
X = np.array([
    [2.0, 1.0, 0.5, 0.1],
    [5.1, 3.1, 2.1, 1.05],
    [5.0, 3.0, 2.0, 1.0],
    [2.1, 0.9, 0.55, 0.15],
    [8.0, 5.0, 4.0, 2.0],
    [8.2, 5.2, 4.2, 2.1],
], dtype=np.float64)
y = np.array([0, 1, 1, 0, 2, 2], dtype=np.int64)
f_sym, params = gen_nn(X, y, hidden_dims=[2,])
ou.newton_quasi.bfgs(f_sym, params, np.ones(len(params)).tolist(), verbose=True, epsilon=1e-2)
# ou.gradient_descent.barzilar_borwein(f_sym, params, np.ones(len(params)).tolist(), verbose=True, epsilon=1e-2) # works well
```

**Comment by moorepants** [comment]
Thanks for the report. We have other open issues on the performance of differentiation you may want to read. Your example is a bit complex for evaluating, it would be best if you can provide a symbolic expression that the Hessian computation is slow for outside of your optimtool code.

**Comment by linjing-lab** [comment]
@moorepants Thanks for your reply. The `gen_nn` function uses classification data to generate symbolic function for `jacobian` computation. The following code can be used to test performance:
```python
import sympy as sp
import numpy as np
X = np.array([
    [2.0, 1.0, 0.5, 0.1],
    [5.1, 3.1, 2.1, 1.05],
    [5.0, 3.0, 2.0, 1.0],
    [2.1, 0.9, 0.55, 0.15],
    [8.0, 5.0, 4.0, 2.0],
    [8.2, 5.2, 4.2, 2.1],
], dtype=np.float64)
y = np.array([0, 1, 1, 0, 2, 2], dtype=np.int64)
f_sym, params = gen_nn(X, y, hidden_dims=[2,])
funcs = sp.Matrix([f_sym])
args = sp.Matrix([params])
res = funcs.jacobian(args)
hes = res.jacobian(args) # hessian computation
```

**Comment by moorepants** [comment]
Maybe this is a duplicate of https://github.com/sympy/sympy/issues/8191

**Comment by linjing-lab** [comment]
I think this is because the function generation code for the classification task is too complex, resulting in slow calculation of the Hessian matrix:
```python
...
if task == 'classification':
    if len(y_data.shape) == 1:
        n_outputs = len(np.unique(y_data))
        y_one_hot = np.zeros((n_samples, n_outputs))
        for i, label in enumerate(y_data):
            y_one_hot[i, int(label)] = 1
    else:
        n_outputs = y_data.shape[1]
        y_one_hot = y_data
    y_processed = y_one_hot
...
if task == 'classification':
    for s in range(n_samples):
        logits = forward(X_data[s].tolist())
        y_true = y_processed[s].tolist()
        log_sum_exp = sp.log(sum(sp.exp(l) for l in logits) + epsilon)
        probs = [sp.exp(l - log_sum_exp) for l in logits]
        for c in range(n_outputs):
            f_sym += -y_true[c] * sp.log(probs[c] + epsilon) # Softmax
    f_sym /= n_samples 
```
Training runs well in second-order computation when turn to regression mode:
```python
# regression
def gen_reg(n_samples=20, seed=0):
    np.random.seed(seed)
    X = np.linspace(-5, 5, n_samples).reshape(-1, 1)
    noise = 0.2 * np.random.randn(n_samples, 3)
    y1 = np.sin(X)
    y2 = 0.5 * X
    y3 = np.cos(X) + 0.2 * X
    y = np.hstack([y1, y2, y3]) + noise
    return X, y

X, y = gen_reg(n_samples=5)
f_sym, params = gen_nn(X, y, hidden_dims=[2,], task='regression')
ou.newton_quasi.bfgs(f_sym, params, np.ones(len(params)).tolist(), verbose=True, epsilon=1e-4)
```
So the overall conclusion is that if the symbol function calculation method of the classification mode can be optimized, the calculation speed of Hessian may be improved, but the overall reason is that the computational complexity of Hessian is too high.

**Comment by SaxdaAnhalta** [comment]
I reproduced the problem on the current SymPy master (1.15.0.dev) under Python 3.12 (Windows).

**Test results & analysis:**

Rebuilding the loss function from the issue (19 parameters, softmax cross-entropy over 6 samples, `hidden_dims=[2]`) gives the following measurements:

- Building `f_sym`: ~0.1 s (552 operations via `count_ops`)
- Gradient (19 derivatives): ~0.7 s (33,327 operations in total — that is a ~60x growth in expression size)
- Hessian via applying `jacobian` twice (19x19 = 361 derivatives): ~70 s

The measurements clearly show that the runtime scales almost linearly with the number of computed rows of the Hessian (2 rows take about 6.5 s, all 19 rows about 70 s). So there is no single bottleneck "hotspot" in the code — the main reason for the long runtime is the extreme growth of the symbolic expressions (expression swell) under repeated differentiation.

**Practical tip as a workaround:**

To compute the Hessian, you should use the function `sympy.hessian(f, params)` directly instead of applying `jacobian()` twice. It exploits the symmetry of the Hessian ($H_{ij} = H_{ji}$) and thereby saves about half of the symbolic derivatives, which noticeably reduces the computation time.

(Transparency note: the benchmark measurements were carried out and verified with the assistance of Claude Code, but I also checked them myself.)

**Comment by linjing-lab** [comment]
@SaxdaAnhalta Thank you for your reply. I have replaced Hessian computations in https://github.com/linjing-lab/optimtool/commit/7e9b8a7554830f338a4a2c5a1c16eca5bcd1687e with `sp.hessian(funcs, args)`. However, as you pointed out, this stems from the complexity of the expressions in classification. I believe this is also why we rely on gradient-based methods for classification.
